server/src/main/java/io/druid/segment/realtime/appenderator/AppenderatorImpl.java

```diff
@@ -240,19 +246,51 @@ public AppenderatorAddResult add(
     final int numAddedRows = sinkRowsInMemoryAfterAdd - sinkRowsInMemoryBeforeAdd;
     rowsCurrentlyInMemory.addAndGet(numAddedRows);
     totalRows.addAndGet(numAddedRows);
+    bytesCurrentlyInMemory.addAndGet(bytesInMemoryAfterAdd - bytesInMemoryBeforeAdd);
 
     boolean isPersistRequired = false;
-    if (!sink.canAppendRow()
-        || System.currentTimeMillis() > nextFlush
-        || rowsCurrentlyInMemory.get() >= tuningConfig.getMaxRowsInMemory()) {
+    boolean persist = false;
+    List<String> persistReasons = Lists.newArrayList();
+
+    if (!sink.canAppendRow()) {
+      persist = true;
+      persistReasons.add("No more rows can be appended to sink");
+    }
+    if (System.currentTimeMillis() > nextFlush) {
```

Comment:
I wonder why we need to check all conditions every time rather than skipping the subsequent checks if `persist` is already true. Is this to log the status of Appenderator?

Reply:
Yes, it's for logging, Gian had a comment earlier about that.